File: appli-android/app/src/main/java/com/example/sportapp/core/data/model/MealEntry.kt
package com.example.sportapp.core.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.example.sportapp.core.utils.CustomDateUtils.getNowISO8601

/**
 * Nutrition A1 (2026-06-17) — portage Android du modèle serveur `meal_entries`
 * (cf. serveur/app/models/meal_entry.py). Table centrale : une consommation dans
 * un repas. Ownership indirect : MealEntry -> Meal -> User (pas de `user_id`).
 *
 * Snapshot D5 : macros per-100g + display_name FIGÉS au moment de l'ajout
 * (historique immuable). FK food/recipe en SET NULL (l'entry survit à la
 * suppression de sa source grâce au snapshot) ; FK meal en CASCADE.
 * Totaux dérivés côté client : total = per_100g x quantity_g / 100.
 */
@Entity(
    tableName = "meal_entries",
    indices = [
        Index(value = ["uuid"], unique = true),
        Index(value = ["meal_uuid"]),
        Index(value = ["food_uuid"]),
        Index(value = ["recipe_uuid"]),
    ],
    foreignKeys = [
        ForeignKey(
            entity = Meal::class,
            parentColumns = ["uuid"],
            childColumns = ["meal_uuid"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = Food::class,
            parentColumns = ["uuid"],
            childColumns = ["food_uuid"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = Recipe::class,
            parentColumns = ["uuid"],
            childColumns = ["recipe_uuid"],
            onDelete = ForeignKey.SET_NULL
        )
    ]
)
data class MealEntry(
    @PrimaryKey
    @ColumnInfo(name = "uuid") val uuid: String,

    @ColumnInfo(name = "meal_uuid") val mealUUID: String,
    @ColumnInfo(name = "food_uuid") val foodUUID: String? = null,       // ref informative (SET NULL)
    @ColumnInfo(name = "recipe_uuid") val recipeUUID: String? = null,   // ref informative (SET NULL)

    @ColumnInfo(name = "display_name") val displayName: String,         // snapshot du nom
    @ColumnInfo(name = "quantity_g") val quantityG: Float,
    @ColumnInfo(name = "portion_label") val portionLabel: String? = null,

    // Snapshot macros per-100g (D5)
    @ColumnInfo(name = "kcal_per_100g") val kcalPer100g: Float,
    @ColumnInfo(name = "protein_per_100g") val proteinPer100g: Float,
    @ColumnInfo(name = "carbs_per_100g") val carbsPer100g: Float,
    @ColumnInfo(name = "fat_per_100g") val fatPer100g: Float,
    @ColumnInfo(name = "fiber_per_100g") val fiberPer100g: Float? = null,
    @ColumnInfo(name = "sugar_per_100g") val sugarPer100g: Float? = null,
    @ColumnInfo(name = "sat_fat_per_100g") val satFatPer100g: Float? = null,
    @ColumnInfo(name = "salt_per_100g") val saltPer100g: Float? = null,
    // Snapshot vitamines & minéraux (pack essentiel ~10, D11 étendu)
    @ColumnInfo(name = "iron_per_100g") val ironPer100g: Float? = null,
    @ColumnInfo(name = "calcium_per_100g") val calciumPer100g: Float? = null,
    @ColumnInfo(name = "magnesium_per_100g") val magnesiumPer100g: Float? = null,
    @ColumnInfo(name = "zinc_per_100g") val zincPer100g: Float? = null,
    @ColumnInfo(name = "potassium_per_100g") val potassiumPer100g: Float? = null,
    @ColumnInfo(name = "sodium_per_100g") val sodiumPer100g: Float? = null,
    @ColumnInfo(name = "vitamin_c_per_100g") val vitaminCPer100g: Float? = null,
    @ColumnInfo(name = "vitamin_d_per_100g") val vitaminDPer100g: Float? = null,
    @ColumnInfo(name = "vitamin_b12_per_100g") val vitaminB12Per100g: Float? = null,
    @ColumnInfo(name = "vitamin_a_per_100g") val vitaminAPer100g: Float? = null,

    @ColumnInfo(defaultValue = "0") val synced: Boolean = false,
    @ColumnInfo(defaultValue = "0") val pendingDeletion: Boolean = false,

    @ColumnInfo(name = "updated_at") val updatedAt: String? = getNowISO8601(),
)

File: serveur/app/models/meal_entry.py
# app/models/meal_entry.py
#
# Nutrition V1 (2026-06-12, cf. docs/NUTRITION_DESIGN.md §3.6) — table centrale.
# Une consommation dans un repas. Ownership indirect : MealEntry -> Meal -> User.
#
# Snapshot D5 : les macros per-100g + display_name sont FIGES au moment de
# l'ajout — l'historique est immuable, corriger/supprimer un aliment du
# catalogue ne reecrit pas le passe. FK food/recipe en SET NULL : l'entry
# survit a la suppression de sa source grace au snapshot.
# Totaux derives cote client : total = per_100g x quantity_g / 100.

from sqlalchemy import Column, Integer, String, Float, ForeignKey, DateTime
from app.database import Base
import uuid


class MealEntry(Base):
    __tablename__ = "meal_entries"

    id = Column(Integer, primary_key=True, index=True)
    uuid = Column(String, nullable=False, unique=True, default=lambda: str(uuid.uuid4()))
    meal_uuid = Column(String, ForeignKey("meals.uuid", ondelete="CASCADE"), nullable=False, index=True)
    # References informatives (SET NULL : le snapshot rend l'entry autonome)
    food_uuid = Column(String, ForeignKey("foods.uuid", ondelete="SET NULL"), nullable=True, index=True)
    recipe_uuid = Column(String, ForeignKey("recipes.uuid", ondelete="SET NULL"), nullable=True, index=True)

    display_name = Column(String, nullable=False)      # snapshot du nom (lisible meme si Food supprime/renomme)
    quantity_g = Column(Float, nullable=False)         # quantite consommee en g
    portion_label = Column(String, nullable=True)      # snapshot du label de portion (« 2 oeufs »)

    # Snapshot macros per-100g (D5)
    kcal_per_100g = Column(Float, nullable=False)
    protein_per_100g = Column(Float, nullable=False)
    carbs_per_100g = Column(Float, nullable=False)
    fat_per_100g = Column(Float, nullable=False)
    # Snapshot micro-nutriments optionnels (D11)
    fiber_per_100g = Column(Float, nullable=True)
    sugar_per_100g = Column(Float, nullable=True)
    sat_fat_per_100g = Column(Float, nullable=True)
    salt_per_100g = Column(Float, nullable=True)
    # Snapshot vitamines & mineraux (pack essentiel ~10, D11 etendu 2026-06-13).
    iron_per_100g = Column(Float, nullable=True)         # mg
    calcium_per_100g = Column(Float, nullable=True)      # mg
    magnesium_per_100g = Column(Float, nullable=True)    # mg
    zinc_per_100g = Column(Float, nullable=True)         # mg
    potassium_per_100g = Column(Float, nullable=True)    # mg
    sodium_per_100g = Column(Float, nullable=True)       # mg
    vitamin_c_per_100g = Column(Float, nullable=True)    # mg
    vitamin_d_per_100g = Column(Float, nullable=True)    # µg
    vitamin_b12_per_100g = Column(Float, nullable=True)  # µg
    vitamin_a_per_100g = Column(Float, nullable=True)    # µg RAE

    updated_at = Column(DateTime(timezone=True), nullable=True)

    def __repr__(self):
        return (
            f"MealEntry(id={self.id}, uuid='{self.uuid}', meal_uuid='{self.meal_uuid}', "
            f"display_name='{self.display_name}', quantity_g={self.quantity_g})"
        )
